File: Algoritm by CPP/Algoritm by CPP/Stack/4949.cpp
//
//  main.cpp
//  Algoritm by CPP
//
//  Created by zaehorang on 2024/02/25.
//  https://www.acmicpc.net/problem/4949

#include <bits/stdc++.h>
typedef long long ll;
using namespace std;
string s;
ll ans = 0;
stack<char> st;
int main(void){
  ios::sync_with_stdio(0);
  cin.tie(0);
  cin >> s;
  int sz = s.length();
  for (int i = 0; i < sz; i++) {
    if (s[i]=='(')
      st.push(s[i]);
    else {
      if (s[i-1] == '(') { // 레이저일 경우
        st.pop(); // 앞에서 막대라고 착각하고 stack에 s[i]를 넣었으므로 pop
        ans+=st.size(); // 막대의 개수만큼 ans에 추가
      }
      else { // 막대의 끝일 경우
        st.pop();  // 막대의 개수를 1 감소
        ans++; // 막대 1개가 절단된 것과 동일한 상황이므로 ans에 1 추가
      }
    }
  }
  cout << ans << "\n";
  return 0;
}
int main(void) {
    ios::sync_with_stdio(0);
    cin.tie(0);
    
    while(true){
        string a;
        getline(cin, a);
        if(a == ".") break;
        
        stack<char> s;
        bool isValid = true;
        
        for (auto c : a){
            if(c == '(' || c == '[') s.push(c);
            else if(c == ')'){
                if(s.empty() || s.top() != '('){ 
                    isValid = false;
                    break;
                }
                
                s.pop();
            }
            else if(c == ']'){
                if(s.empty() || s.top() != '['){
                    isValid = false;
                    break;
                }
                s.pop();
            }
            
        }
        
        if(isValid && s.empty()){
            cout << "yes" << '\n';
        } else cout << "no\n";
    }
}

File: Algorithm by Swift/Algorithm by Swift/Stack/4949.swift
//
//  main.swift
//  Algorithm by Swift
//
//  Created by zaehorang on 2024/02/25.
//  https://www.acmicpc.net/problem/4949

//MARK: - 처음 답안 코드
//while let str = readLine(), str != "." {
//    
//    var s = [Character]()  // stack
//    
//    for char in str {
//        
//        if char == "[" || char == "(" {
//            s.append(char)
//        } else if let last = s.last {
//            if (last == "[" && char == ")") || (last == "(" && char == "]") { // 닫는 괄호와 스택의 top 괄호가 짝이  n아니라면
//                break
//            } else if last == "[" && char == "]" {  // 짝이 맞다면
//                s.removeLast()
//            } else if  last == "(" && char == ")" {
//                s.removeLast()
//            }
//        } else if char == "]" || char == ")" {  // 닫는 괄호가 먼저 나오면
//            s.append(char)  //  스택에 push하고 반복문 종료
//            break
//        }
//    }
//    
//    // 스택에 괄호 유무 확인
//    if !s.isEmpty {
//        print("no")
//    } else {
//        print("yes")
//    }
//
//}

//MARK: - 더 짧게 구성

while let str = readLine(), str != "."  {
    
    var s = [Character]()  // stack
    var isValid = true
    
    for a in str {
        if a == "(" || a == "[" {
            s.append(a)
        } else if a == ")" {
            if s.isEmpty || s.last != "(" {
                isValid = false
                break
            } else {
                s.removeLast()
            }
                
        } else if a == "]" {
            if s.isEmpty || s.last != "[" {
                isValid = false
                break
            } else {
                s.removeLast()
            }
        }
    }
    
    if s.isEmpty && isValid {
        print("yes")
    } else {
        print("no")
    }
    
}
